ariadne/utils.py

```diff
@@ -0,0 +1,260 @@
+    second_station = df[df['station'] == 1][['r', 'phi', 'z', 'track']]
+    second_station.columns = ['r_right', 'phi_right', 'z_right', 'track_right']
+    rows = itertools.product(first_station.iterrows(), second_station.iterrows())
+    df = pd.DataFrame(left.append(right) for (_, left), (_, right) in rows)
+    df_fakes = df[(df['track_left'] == -1) & (df['track_right'] == -1)]
+    df = df[(df['track_left'] != df['track_right'])]
+    df = pd.concat([df, df_fakes], axis=0)
+    df = df.sample(frac=1).reset_index(drop=True)
+    return df.reset_index(drop=True)
+
+def weights_update(model, checkpoint):
+    model_dict = model.state_dict()
+    pretrained_dict = checkpoint['state_dict']
+    real_dict = {}
+    for (k, v) in model_dict.items():
+        needed_key = None
+        for pretr_key in pretrained_dict:
+            if k in pretr_key:
+                needed_key = pretr_key
+                break
+        assert needed_key is not None, "key %s not in pretrained_dict %r!" % (k, pretrained_dict.keys())
+        real_dict[k] = pretrained_dict[needed_key]
+
+    model.load_state_dict(real_dict)
+    model.eval()
+    return model
+
+def load_data(input_dir, file_mask, n_samples):
+    flag = 0
+    files = []
+    data_merged = {}
+    datafiles = glob.glob(f'{input_dir}/{file_mask}')
+    for f in datafiles:
+        one_file_data = np.load(f, mmap_mode='r', allow_pickle=False)
+        first_file_len = len(one_file_data[one_file_data.files[0]])
+        files = one_file_data.files
+        lengths = [len(one_file_data[f]) == first_file_len for f in files]
+        assert all(lengths), 'Lengths of files in npz are not equal!'
+        if flag == 0:
+            data_merged = dict(one_file_data.items())
+            flag = 1
+        else:
+            data_merged = {k: np.concatenate((data_merged[k], i), 0) for k, i in one_file_data.items()}
+    if n_samples is None:
+        n_samples = len(data_merged[files[0]])
+    else:
+        n_samples = min(n_samples, len(data_merged[files[0]]))
+    return {key: item[:n_samples] for key, item in data_merged.items()}
+
+
+def find_nearest_hit_no_faiss(ellipses, last_station_hits, return_numpy=False):
+    centers = ellipses[:, :2]
+    dists = torch.cdist(last_station_hits.float(), centers.float())
+    minimal = last_station_hits[torch.argmin(dists, dim=0)]
+    is_in_ellipse = point_in_ellipse(ellipses, minimal)
+    if return_numpy:
+        minimal = minimal.detach().cpu().numpy()
+        is_in_ellipse = is_in_ellipse.detach().cpu().numpy()
+    return minimal, is_in_ellipse
+
+def find_nearest_hit(ellipses, last_station_hits):
+    #numpy, numpy -> numpy, numpy
+    index = faiss.IndexFlatL2(2)
+    index.train(last_station_hits.astype('float32'))
+    index.add(last_station_hits.astype('float32'))
+
+    #ellipses = torch_ellipses.detach().cpu().numpy()
+    centers = ellipses[:, :2]
+    d, i = index.search(np.ascontiguousarray(centers.astype('float32')), 1)
+    d = np.sqrt(d)
```

Comment:
why we need this var?

Reply:
`_, i = index.search(np.ascontiguousarray(centers.astype('float32')), 1)`

Reply:
don't need, it's artifact of outdated version